src/ast/recfun_decl_plugin.h

```diff
@@ -0,0 +1,264 @@
+            def& get_def(symbol const& s) { return *(m_defs[s]); }
+            bool has_case_def(const symbol& s) const { return m_case_defs.contains(s); }
+            case_def& get_case_def(symbol const& s) { return *(m_case_defs[s]); }
+            bool is_declared(symbol const& s) const { return m_defs.contains(s); }
+
+            func_decl * get_decl(def const & d);
+            func_decl * get_decl(case_pred const & d);
+        private:
+            func_decl * mk_fun_pred_decl(unsigned num_parameters, parameter const * parameters, 
+                                         unsigned arity, sort * const * domain, sort * range);
+            func_decl * mk_fun_defined_decl(decl_kind k,
+                                            unsigned num_parameters, parameter const * parameters, 
+                                            unsigned arity, sort * const * domain, sort * range);
+        };
+    }
+
+    // Varus utils for recursive functions
+    class util {
+        friend class decl::plugin;
+        typedef map<symbol, func_decl*, symbol_hash_proc, symbol_eq_proc> func_decl_map;
+        
+        ast_manager &           m_manager;
+        family_id               m_family_id;
+        th_rewriter             m_th_rw;
+        decl::plugin *          m_plugin;
+        func_decl_map           m_func_decls;
+
+        bool compute_is_immediate(expr * rhs);
+    public:
+        util(ast_manager &m);
+        ~util();
+
+        ast_manager & m() { return m_manager; }
+        bool is_case_pred(app * e) const { return is_app_of(e, m_family_id, OP_FUN_CASE_PRED); }
+        bool is_defined(app * e) const { return is_app_of(e, m_family_id, OP_FUN_DEFINED); }
+        bool is_macro(app * e) const { return is_app_of(e, m_family_id, OP_FUN_MACRO); }
+        bool owns_app(app * e) const { return e->get_family_id() == m_family_id; }
+        
+        bool is_defined_or_macro(app * e) const {
+            return is_app_of(e, m_family_id, OP_FUN_DEFINED) ||
+                   is_app_of(e, m_family_id, OP_FUN_MACRO);
+        }
+
+        //<! add a function declaration
+        def * decl_fun(symbol const & s, sort * args, unsigned n_args, sort * range);
+
+        //<! Add a function definition (TODO: allow mutual definitions)
+        def * add_fun(symbol const & s, sort * args, unsigned n_args, sort * range, var * vars, unsigned n_vars, expr * rhs);
+
+        def& get_def(symbol const & s) {
+            SASSERT(m_plugin->has_def(s));
+            return m_plugin->get_def(s);
+        }
+
+        case_def& get_case_def(symbol const & s) {
+            SASSERT(m_plugin->has_case_def(s));
+            return m_plugin->get_case_def(s);
+        }
+
+        app* mk_fun_defined(def const & d, unsigned n_args, expr * const * args) {
+            func_decl * fd = 0;
+            auto& name = d.get_name();
+            if (! m_func_decls.find(name, fd)) {
+                fd = m_plugin->get_decl(d);
+                m_func_decls[name] = fd;
```

Comment:
m_func_decls[name] = fd is probably not going to work as intended. This could perhaps be fixed in our hash-tables, but m_func_decls.insert(name, fd) works too, or insert_if_not_there does the job in a single call.